compiler/src/parsing/ast_helper.re

```diff
@@ -384,25 +390,111 @@ module Expression = {
+        if (List.exists(
+              expr =>
+                switch (expr) {
+                | ListSpread(_) => true
+                | ListItem(_) => false
+                },
+              rest,
+            )) {
+          collection_concat(
+            ~loc,
+            ~attributes?,
+            PExpListConcat,
+            List.map(
+              expr =>
+                switch (expr) {
+                | ListSpread(expr, loc) => (
+                    PExpSpreadExpr,
+                    {...expr, pexp_loc: loc},
+                  )
+                | ListItem(expr) => (
+                    PExpNonSpreadExpr,
+                    // Still convert to a single-element list to make later compilation steps easier
+                    tuple_construct(
+                      ~loc=expr.pexp_loc,
+                      ~attributes?,
+                      ident_cons,
+                      [
+                        expr,
+                        tuple_construct(~loc=expr.pexp_loc, ident_empty, []),
+                      ],
+                    ),
+                  )
+                },
+              a,
+            ),
+          );
+        } else {
+          List.fold_left(
+            (acc, expr) => {
+              switch (expr) {
+              | ListItem(expr) =>
+                tuple_construct(~loc, ~attributes?, ident_cons, [expr, acc])
+              | ListSpread(_, loc) =>
+                failwith(
+                  "Impossible: non-final list spread when existence has been disproven",
+                )
+              }
+            },
+            base,
+            rest,
+          );
+        };
+      };
+    {...list, pexp_loc: loc};
+  };
+  let array_items = (~loc, ~attributes=?, a) => {
+    let no_spreads =
+      List.for_all(
+        x => {
+          switch (x) {
+          | ArrayItem(_) => true
+          | ArraySpread(_) => false
+          }
+        },
+        a,
+      );
+    if (no_spreads) {
+      array(
+        ~loc,
+        ~attributes?,
+        List.map(
+          expr =>
             switch (expr) {
-            | ListItem(expr) =>
-              tuple_construct(~loc, ~attributes?, ident_cons, [expr, acc])
-            | ListSpread(_, loc) =>
-              raise(
-                SyntaxError(
-                  loc,
-                  "A list spread can only appear at the end of a list.",
-                ),
+            | ArraySpread(_) =>
+              failwith(
+                "Impossible: spread in array when existence has been disproven",
+              )
+            | ArrayItem(expr) => expr
+            },
+          a,
+        ),
+      );
+    } else {
+      collection_concat(
+        ~loc,
+        ~attributes?,
+        PExpArrayConcat,
+        List.map(
+          x => {
+            switch (x) {
+            | ArrayItem(expr) => (
+                PExpNonSpreadExpr,
+                // Still convert to a single-element array to make later compilation steps easier
```

Comment:
Similar thing here.

Reply:
Wouldn't it not matter much for arrays since the efficiency of spreads at the end is only relevant for linked lists?

Reply:
Just reducing the overall number of allocations. They're pretty expensive, especially once you factor in the GC overhead.

Reply:
Just to clarify, you mean to say that specifically single elements should be grouped with the next spread in the same way that you suggested for lists (like treating `[> a, b, ...c, d, e, ...f]` as `[> [> a, b, ...c], [> d, e, ...f]]`)?

Reply:
In this case, it would be `[> [> a, b], c, [> d, e] f]`.

Reply:
I see, that makes more sense